source/common/http/async_client_utility.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "envoy/http/async_client.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Keeps track of active async HTTP requests to be able to cancel them on destruction.
+ */
+class AsyncClientRequestTracker {
+public:
+  /**
+   * Cancels all known active async HTTP requests.
+   */
+  ~AsyncClientRequestTracker();
+  /**
+   * Includes a given async HTTP request into a set of known active requests.
+   * @param request request handle
+   */
+  void add(AsyncClient::Request& request);
+  /**
+   * Excludes a given async HTTP request from a set of known active requests.
+   * @param request request handle
+   */
+  void remove(const AsyncClient::Request& request);
```

Comment:
It doesn't make sense to pass a non-const object above and a const object here. I would just pass a non-const in both places and remove the const_cast. Thank you!

/wait

Reply:
`AsyncClientRequestTracker::remove()` is intended for use in the context of `AsyncClient::Callbacks::onSuccess(request, response)` and `AsyncClient::Callbacks::onFailure(request, reason)`.

Notice that `onSuccess / onFailure` methods will now have `request` as an extra parameter to enable correlation between requests and responses.

The proposed interface for `AsyncClient::Callbacks` looks like this:

```c++
  /**
   * Notifies caller of async HTTP request status.
   *
   * To support a use case where a caller makes multiple requests in parallel,
   * individual callback methods provide request context corresponding to that response.
   */
  class Callbacks {
  public:
    virtual ~Callbacks() = default;

    /**
     * Called when the async HTTP request succeeds.
     * @param request  request handle.
     *                 NOTE: request handle is passed for correlation purposes only, e.g.
     *                 for client code to be able to exclude that handle from a list of
     *                 requests in progress.
     * @param response the HTTP response
     */
    virtual void onSuccess(const Request& request, ResponseMessagePtr&& response) PURE;

    /**
     * Called when the async HTTP request fails.
     * @param request request handle.
     *                NOTE: request handle is passed for correlation purposes only, e.g.
     *                for client code to be able to exclude that handle from a list of
     *                requests in progress.
     * @param reason  failure reason
     */
    virtual void onFailure(const Request& request, FailureReason reason) PURE;
  };
```

`request` was made `const Request&` to prevent any unintended usage, e.g. calling `request.cancel()`.

So, `const Request&` in `Callbacks` causes the same in `AsyncClientRequestTracker::remove()` (otherwise, `const_cast` will be a part of every call to `AsyncClientRequestTracker::remove()`).

Do you want me to remove `const AsyncClient::Request&` from both `AsyncClientRequestTracker` and `Callbacks` ?

Reply:
Hmm OK. Can you add a comment above the const_cast as to why it's required? Thank you.

Reply:
added comments